### ISSUE TITLE
DPMI problems on hosteurope server with -I "cpu_vm_dpmi native"

### ISSUE THREAD

**Issue by ecm-pushbx** [dosemu2/dosemu2#1615]
**Describe the bug**

Running on this server (ulukai.org or pushbx.org) with `-I "cpu_vm_dpmi native"` causes some problems. I described the server some in the D/B bit discussion: https://github.com/dosemu2/dosemu2/discussions/1609

With the latest devel commit, unpatched, I get this error upon attempting to enter Protected Mode:

```
ERROR: Your kernel does not support UC_STRICT_RESTORE_SS and the work-around in dosemu is not enabled.
```

Here I'm linking [the full output text](https://pushbx.org/ecm/test/20220414/1427.txt) and [the boot.log](https://pushbx.org/ecm/test/20220414/1427.log).
 
So now I'm using the latest devel commit with one change:

```
~/proj/newnewdosemu/20220329$ git diff
diff --git a/src/include/emu.h b/src/include/emu.h
index 4d49376c2..0460f6a5c 100644
--- a/src/include/emu.h
+++ b/src/include/emu.h
@@ -497,6 +497,6 @@ extern int hlt_unregister_handler_vm86(Bit16u start_addr);

 extern const char *Path_cdrom[];

-#define WANT_SIGRETURN_WA 0
+#define WANT_SIGRETURN_WA 1

 #endif /* EMU_H */
~/proj/newnewdosemu/20220329$ git describe
2.0pre9-dev-119-g99464dbf8
~/proj/newnewdosemu/20220329$
```

This works fine when just running (not tracing) a DPMI client. However, when tracing then at an apparently random point dosemu crashes:

```
ERROR: Fault in dosemu code, in_dpmi=1
ERROR: cpu exception in dosemu code outside of DPMI client!
sig: 11 trapno: 0x0e  errorcode: 0x00000004  cr2: 0x9a159357
eip: 0x55c05192806f  esp: 0x7ffdcfd44b40  eflags: 0x00010213
cs: 0x0033  ds: 0x0000  es: 0x0000  ss: 0x002b
fs: 0x0000  gs: 0x0000
fsbase: 0x7f6c8ccd67c0 gsbase: (nil)

ERROR: Please install gdb, update dosemu from git, compile it with debug
info and make a bug report with the content of ~/.dosemu/boot.log at
https://github.com/dosemu2/dosemu2/issues
Please provide any additional info you can, like the test-cases,
URLs and all the rest that fits.

==1605573==WARNING: ASan is ignoring requested __asan_handle_no_return: stack top: 0x000000000000; bottom 0x7ffdcfd43000; size: 0xffff8002302bd000 (-140728090243072)
False positive error reports may follow
For details see https://github.com/google/sanitizers/issues/189
ERROR: Coopth: not all detached threads properly shut down
ERROR:  tid=269 state=1 name="msdos ext thr" off=0x7
Terminated with signal 11
dpmitest$
```

Here is [the full output text](https://pushbx.org/ecm/test/20220414/1433.txt) and [the boot.log](https://pushbx.org/ecm/test/20220414/1433.log).

**To Reproduce**

First bug: Just try to enter Protected Mode.

Second bug: Enter Protected Mode with lDebugX then trace/proceed through the PM code using T/TP/P/G commands. Continue for several dozen steps and a crash will happen.

**Attach program/game binaries or provide an URL**
Executables provided in https://pushbx.org/ecm/test/20220414/

**Comment by stsp** [comment]
> ERROR: Your kernel does not support UC_STRICT_RESTORE_SS and the work-around in dosemu is not enabled.

So can you check that adb226f4f
makes this exit clean, rather than
the crash you showed before?

Please open another ticket for that
"stack patch failure" error.

Please provide the -D9+M log for
the new DPMI crash.

**Comment by ecm-pushbx** [comment]
> > ERROR: Your kernel does not support UC_STRICT_RESTORE_SS and the work-around in dosemu is not enabled.
> 
> So can you check that [adb226f](https://github.com/dosemu2/dosemu2/commit/adb226f4f7354705fbcac5949891b1f8e8abf3cf) makes this exit clean, rather than the crash you showed before?

Do you mean I should build the commit prior to this one, without the `WANT_SIGRETURN_WA 1` edit, and compare the crash behaviour?

**Comment by stsp** [comment]
Yes, or just revert that commit and rebuild.

**Comment by ecm-pushbx** [comment]
> Please provide the -D9+M log for the new DPMI crash.

Here's [the output](https://pushbx.org/ecm/test/20220414/1511.txt) and [the boot.log](https://pushbx.org/ecm/test/20220414/1511.log) (it is larger than 19 MB).

**Comment by ecm-pushbx** [comment]
[Output](https://pushbx.org/ecm/test/20220414/1523.txt) and [the boot.log](https://pushbx.org/ecm/test/20220414/1523.log) of trying to enter Protected Mode using 84a8625f858c75081094ad7e3bd6694a20458e42.

**Comment by ecm-pushbx** [comment]
> Please open another ticket for that "stack patch failure" error.

https://github.com/dosemu2/dosemu2/issues/1616

**Comment by stsp** [comment]
Thanks, so this confirms that the
shutdown is now clean, w/o any
extra messages, right?

**Comment by ecm-pushbx** [comment]
Yes.

**Comment by stsp** [comment]
Good to know.
The second crash you have
is because your kernel doesn't
seem to support CONFIG_X86_ESPFIX64.

[stk2.c.txt](https://github.com/dosemu2/dosemu2/files/8489924/stk2.c.txt)
Here is the test-case.

**Comment by ecm-pushbx** [comment]
> Good to know. The second crash you have is because your kernel doesn't seem to support CONFIG_X86_ESPFIX64.
> 
> [stk2.c.txt](https://github.com/dosemu2/dosemu2/files/8489924/stk2.c.txt) Here is the test-case.

Compiling this with `gcc` and running it gives:

```
./stk2
ss=0x2b, new_ss=0x2f
In sighandler: esp=7ffe5a349390 trapno=0xd cr2=0 eip=5566adbfb47c
old_esp=0x7ffe5a349390, new_esp=0x7ffe5a349390
No bug here.
```

**Comment by stsp** [comment]
Well, it is also quite obvious
from the log that it doesn't happen
always, but rather very infrequently.
In a test_case try to add `exit(1);`
after printf("BUG") and then run the
test in a loop:
`while ./stk2 ; do true; done`
or something like that.
If it eventually trigger the bug, it
will exit the loop.

**Comment by ecm-pushbx** [comment]
> Well, it is also quite obvious from the log that it doesn't happen always, but rather very infrequently. In a test_case try to add `exit(1);` after printf("BUG") and then run the test in a loop: `while ./stk2 ; do true; done` or something like that. If it eventually trigger the bug, it will exit the loop.

It's been running in a loop like this since Friday 17:30 (+0200 timezone). It has yet to fail. Perhaps the test case just doesn't work well.

**Comment by stsp** [comment]
Another guess is that your
debugger also needs a espfix
technique, as you say the
problem is only happening
when tracing.
If your debugger ever does
iret to the tracee and that
switches from 32 to 16bit
stack, then you need to
implement the espfix technique
yourself.

**Comment by stsp** [comment]
```
DPMI: int 0x2f
DPMI: call inthandler 0x2f(0x1680) at 0x97:0x000195
                ret=0x107:0x00c08e
DPMI: cs=0x97, ss=0xff, ds=0xff, es=0xff, fs=0, gs=0 ip=0x195
DPMI: switch to dpmi
DPMI: Return to client at 0097:00000195, Stack 0xff:0x0000c33c, flags=0x10246
DPMI: switch to dosemu
DPMI: cs=0x97, ss=0xff, ds=0xff, es=0xff, fs=0, gs=0 ip=0x195
DPMI: Starting MSDOS pm call
eip: 0x00000196  esp: 0x8f30c33c  eflags: 0x00010246
        trapno: 0x0d  errorcode: 0x00000000  cr2: 0x41afdfe0
        cs: 0x0097  ds: 0x00ff  es: 0x00ff  ss: 0x00ff  fs: 0x0000  gs: 0x0000
EAX: 00001680  EBX: 00000001  ECX: 00000001  EDX: 0000bf3e
ESI: 0000824c  EDI: 0000b732  EBP: 0000c36a
OPS  : f4 f4 f4 f4 f4 f4 f4 f4 f4 f4 -> f4 f4 f4 f4 f4 f4 f4 f4 f4 f4
STACK: <invalid memory> SS:ESP points to invalid memory
```

So indeed that happens after
int2f. Likely when your debugger
does iret from that int2f.
Who hooks int2f in your case?

https://github.com/dosemu2/dosemu2/blob/devel/src/dosext/dpmi/dpmi.c#L4625
Here we create the raw iret
frame. Maybe it would be
safer to use trampoline to
save you from the espfix
workout.

**Comment by stsp** [comment]
I created the espfix branch where
I added the possible work-around
for that problem.

**Comment by ecm-pushbx** [comment]
> So indeed that happens after
> int2f. Likely when your debugger
> does iret from that int2f.
> Who hooks int2f in your case?

I don't think the debugger (nor this test case) ever *hooks* Protected Mode interrupt 2Fh. The call appears to be a function 1680h call (Release time slice); the debugger will issue such calls while idle in its wait on the `rawinput` function's call to `getc`. As you can see in the output I enabled the DCO flag 800h, instructing the debugger to use `rawinput` instead of calling interrupt 21h service 0Ah (either in 86M or PM).

I'm compiling your branch now and will report back on whether it improves things.

**Comment by ecm-pushbx** [comment]
> If your debugger ever does
> iret to the tracee and that
> switches from 32 to 16bit
> stack, then you need to
> implement the espfix technique
> yourself.

Interesting, this might happen when your (dosemu's) int 2Fh returns to the debugger. Not sure how it'd get a 32-bit stack but the debugger certainly uses a 16-bit stack for itself.

[Here's the call to 2F.1680](https://hg.pushbx.org/ecm/ldebug/file/release4/source/lineio.asm#l1714) "in the given mode", meaning it will call PM interrupt 2Fh if the debugger is in PM when the `idle` function is called. And [here's the way the debugger obtains its `ss` selector](https://hg.pushbx.org/ecm/ldebug/file/release4/source/pminit.asm#l80). It is expected that this will be a selector with D/B bit = 0 and limit = 0FFFFh.

**Comment by ecm-pushbx** [comment]
It crashed again, though perhaps a bit later. [Here's the output](https://pushbx.org/ecm/test/20220421/2100.txt) and [here the log](https://pushbx.org/ecm/test/20220421/2100.log).

**Comment by ecm-pushbx** [comment]
It does crash after a while when just left at the (`rawinput`) prompt in Protected Mode, no tracing needed.

**Comment by ecm-pushbx** [comment]
Here's the output of `rvm` and `dl` on the given selectors. As expected the limits are 0FFFFh and the D/B bit is clear. 

```
-g 211
Succeeded in unhooking interrupt 2Fh.
AX=0001 BX=00E6 CX=0005 DX=000A SP=FFFC BP=FFFC SI=070E DI=8A76
DS=00B7 ES=00BF SS=00B7 CS=0117 IP=0211 NV UP EI PL NZ NA PO NC
0117:0211 8CCB              mov     bx, cs
#rvm
Code segment=0FB0 selector=0107
Data segment=0378 selector=00FF
Entry segment=0378 selector=00AF
Auxbuff segment=1D10 selector=0127
History segment=1F11 selector=012F
#dl 107
0107 base=0000FB00 limit=0000FFFF attr=00FB
#dl ff
00FF base=00003780 limit=0000FFFF attr=00F3
#dl af
00AF base=00003780 limit=0000FFFF attr=00FB
#dl 127
0127 base=0001D100 limit=0000FFFF attr=00F3
#dl 12f
012F base=0001F110 limit=0000FFFF attr=00F3
#
```

**Comment by ecm-pushbx** [comment]
Sorry, prior log was without `-D9+M`. Here's a crash with that switch, after idling for a while. (It seemed like the crash happened when I pressed a key again after a while.) Output: https://pushbx.org/ecm/test/20220421/2130.txt log: https://pushbx.org/ecm/test/20220421/2130.log

**Comment by ecm-pushbx** [comment]
I also uploaded the `ldebugx.com` file I use currently to that server directory, including its map and listing files.

**Comment by ecm-pushbx** [comment]
I could probably make the debugger use a 32-bit stack on a 386 machine. Do you think that'd help?

**Comment by stsp** [comment]
It will definitely help to see if we are on
a right track.
Does the crash happen in debugger
or debugee? (whose cs is printed in a
crash log?)

**Comment by ecm-pushbx** [comment]
The `cs` is dosemu's I believe. However from the stack (low 16 bits of `sp`) and `ax`  value I am reasonably sure it occurs in the `idle` function. I tried to patch the debugger to clear the high word of `esp` before it calls the interrupt 2Fh function. This did not avoid the crash though, although it may have delayed it:

```
-u fb0:c0cf l 7
0FB0:C0CF F6069D0140        test    byte [019D], 40
0FB0:C0D4 7568              jnz     C13E
-u cs:a0 l 7
2127:00A0 660FB7E4          movzx   esp, sp
2127:00A4 90                nop
2127:00A5 90                nop
2127:00A6 90                nop
-m cs:a0 l 7 fb0:c0cf
```

**Comment by stsp** [comment]
Yes, cs obviously from dosemu, sorry,
whose ss?

**Comment by ecm-pushbx** [comment]
The `ss` is my debugger's own data/stack selector. As an experiment I disabled `rawinput`, entered PM, then wrote a small helper (in the debuggee's `cs`) that sets the B bit of the data/stack selector of the debugger. [Partial output here](https://pushbx.org/ecm/test/20220421/2200.txt). This has not yet crashed.

**Comment by ecm-pushbx** [comment]
I enabled `rawinput` use after running the helper. The debugger has been sitting at its PM prompt, mostly idle,  for longer than 10 minutes now without crashing.

**Comment by ecm-pushbx** [comment]
Found another possible factor. There is a field in the "Real Mode Register Structure" that [is supposed to be zero](http://www.delorie.com/djgpp/doc/dpmi/ch4.6.html). But I didn't set it to zero. [This is the code](https://hg.pushbx.org/ecm/ldebug/file/c880ad9d3cfc/source/debug.asm#l5470). This is the patch:

```
ldebug/source$ hg id
c880ad9d3cfc+ tip
ldebug/source$ hg d
diff --git a/source/debug.asm b/source/debug.asm
--- a/source/debug.asm
+++ b/source/debug.asm
@@ -5468,6 +5468,8 @@
        xchg al, ah
        mov word [bp + ?86m_call_struc +20h], ax        ; flags
        xor ax, ax
+       mov word [bp + ?86m_call_struc +0Ch + 2], ax
+       mov word [bp + ?86m_call_struc +0Ch], ax
        mov word [bp + ?86m_call_struc +2Eh], ax        ; sp
        mov word [bp + ?86m_call_struc +30h], ax        ; ss
        mov ax, word [bp + ?es_value]                   ; usually [pspdbg]
ldebug/source$
```

No, it took longer but it did just crash again now.

**Comment by ecm-pushbx** [comment]
This latest run (lDebugX c880+ with the patch to set field at 0Ch) with -D9+M crashed after longer than usual but did crash. Output: https://pushbx.org/ecm/test/20220421/2250.txt log: https://pushbx.org/ecm/test/20220421/2250.log

**Comment by ecm-pushbx** [comment]
I have a suggestion. The "`Protected-mode state dump`" from the crash currently shows as "`STACK: <invalid memory>`". Can you make it to add a stack dump assuming only the low 16 bits of `esp` should be used? This could help to determine exactly which call led to the crash.

**Comment by stsp** [comment]
Its quite clear that int2f leads
to a crash, but the problem is
not in your code. Corruption
happens when dosemu is
handling its own hlt.
I still think this is a ESPFIX64
problem of a kernel, but I added
the patch that can make things
somewhat better, please check.

**Comment by stsp** [comment]
I also added another possible
mitigation to espfix2 branch.
Please check also that.

**Comment by ecm-pushbx** [comment]
I made a quick and dirty edit to DPMI_show_state to add a STACK16 dump: https://pushbx.org/ecm/test/20220421/2320.diff

**Comment by stsp** [comment]
It is very likely that in espfix2
branch the stack will be printed
as you expect.

**Comment by ecm-pushbx** [comment]
f170b306b doesn't help, it crashes very quickly.

**Comment by ecm-pushbx** [comment]
c85557a0c (espfix2) seems good for at least a moment.

**Comment by stsp** [comment]
Please attach the log of f170b30

**Comment by ecm-pushbx** [comment]
> Please attach the log of [f170b30](https://github.com/dosemu2/dosemu2/commit/f170b306be2123f0d0ab0c6b18fe3872671795f9)

Output is at https://pushbx.org/ecm/test/20220421/2330.txt and log at https://pushbx.org/ecm/test/20220421/2330.log

**Comment by ecm-pushbx** [comment]
> I made a quick and dirty edit to DPMI_show_state to add a STACK16 dump: https://pushbx.org/ecm/test/20220421/2320.diff

Also I lengthened the stack part shown above the current `ss:(e)sp`

**Comment by stsp** [comment]
Hmm, so esp seems corrupted
already when dosemu detects int2f
call, but iret frame successfully
created??
Please do this:
```
diff --git a/src/dosext/dpmi/dpmi.c b/src/dosext/dpmi/dpmi.c
index 42097cfc4..806b5794d 100644
--- a/src/dosext/dpmi/dpmi.c
+++ b/src/dosext/dpmi/dpmi.c
@@ -4602,6 +4602,7 @@ static int dpmi_gpf_simple(sigcontext_t *scp, uint8_t *lina, void *sp, int *rv)
         return 1;
       }
       D_printf("DPMI: int 0x%04x, AX=0x%04x\n", inum, _LWORD(eax));
+D_printf("%s", DPMI_show_state(scp));
 #ifdef USE_MHPDBG
       if (mhpdbg.active) {
         if (dpmi_mhp_intxxtab[inum]) {
```
and attach the log.

**Comment by ecm-pushbx** [comment]
Which revision should I use with that patch?

**Comment by stsp** [comment]
> Also I lengthened the stack part shown above the current ss:(e)sp

Please make a PR for that, if
you want. I can't look into 2 things
at once.

> Which revision should I use with that patch?

Latest devel.

**Comment by stsp** [comment]
Devel updated.

**Comment by ecm-pushbx** [comment]
https://github.com/dosemu2/dosemu2/commit/c85557a0c92353929271c4a9fe18945d34195848 (espfix2) has been running since 23:40 now (20 minutes). No crash.

**Comment by stsp** [comment]
Cool but lets focus on devel for now.

**Comment by ecm-pushbx** [comment]
Building right now.

**Comment by ecm-pushbx** [comment]
What about `WANT_SIGRETURN_WA 1` by the way? Do I need it because of my weird kernel version? Or is there something wrong with dosemu that could be fixed?

**Comment by ecm-pushbx** [comment]
Running g0470b1a59 with your added DPMI_show_state call. Crashes soon as before.

Output: https://pushbx.org/ecm/test/20220421/2400.txt Log: https://pushbx.org/ecm/test/20220421/2400.log

**Comment by ecm-pushbx** [comment]
Latest log has this:

```
DPMI: switch to dosemu
DPMI: Return to dosemu at 0107:0000d179, Stack 0xff:0x0000c340, flags=0x13246
DPMI: cs=0x107, ss=0xff, ds=0xff, es=0xff, fs=0, gs=0 ip=0xd179
DPMI: switch to dpmi
DPMI: Return to client at 0107:0000d179, Stack 0xff:0x0000c340, flags=0x13246
DPMI: switch to dosemu
DPMI: cs=0x107, ss=0xff, ds=0xff, es=0xff, fs=0, gs=0 ip=0xc113
DPMI: int 0x002f, AX=0x1680
eip: 0x0000c113  esp: 0x8ef8c358  eflags: 0x00010246
	trapno: 0x0d  errorcode: 0x0000017a  cr2: 0x408d9640
	cs: 0x0107  ds: 0x00ff  es: 0x00ff  ss: 0x00ff  fs: 0x0000  gs: 0x0000
EAX: 00001680  EBX: 00000001  ECX: 00000001  EDX: 0000bf4e
ESI: 0000824c  EDI: 0000b742  EBP: 0000c37a
OPS  : 0b 06 be 00 07 74 09 b8 80 16 -> cd 2f 84 c0 74 50 e8 ad d0 75 
STACK: <invalid memory> SS:ESP points to invalid memory
DPMI: int 0x2f
DPMI: call inthandler 0x2f(0x1680) at 0x97:0x000195
		ret=0x107:0x00c115
DPMI: more hlt to handle
DPMI: Starting MSDOS pm call
eip: 0x00000196  esp: 0x8ef8c34c  eflags: 0x00010246
	trapno: 0x0d  errorcode: 0x00000000  cr2: 0x408d9640
	cs: 0x0097  ds: 0x00ff  es: 0x00ff  ss: 0x00ff  fs: 0x0000  gs: 0x0000
EAX: 00001680  EBX: 00000001  ECX: 00000001  EDX: 0000bf4e
ESI: 0000824c  EDI: 0000b742  EBP: 0000c37a
OPS  : f4 f4 f4 f4 f4 f4 f4 f4 f4 f4 -> f4 f4 f4 f4 f4 f4 f4 f4 f4 f4 
STACK: <invalid memory> SS:ESP points to invalid memory
GPF on selector 0xff base=00003780 lim=ffff
ERROR: Fault in dosemu code, in_dpmi=1
ERROR: cpu exception in dosemu code outside of DPMI client!
sig: 11 trapno: 0x0e  errorcode: 0x00000004  cr2: 0x99f04f59
eip: 0x55792e99d414  esp: 0x7ffd9f3a30f0  eflags: 0x00010217
cs: 0x0033  ds: 0x0000  es: 0x0000  ss: 0x002b
fs: 0x0000  gs: 0x0000
fsbase: 0x7fec9f3047c0 gsbase: (nil)

ERROR: Please install gdb, update dosemu from git, compile it with debug
info and make a bug report with the content of ~/.dosemu/boot.log at
https://github.com/dosemu2/dosemu2/issues
Please provide any additional info you can, like the test-cases,
URLs and all the rest that fits.
```

0107h:C115h points exactly behind the PM int 2Fh opcode (in the c880ad+ lDebugX build I'm using now).

**Comment by ecm-pushbx** [comment]
I'll get back to you this weekend if you need more testing or anything.

**Comment by stsp** [comment]
> What about WANT_SIGRETURN_WA 1 by the way? Do I need it because of my weird kernel version?

Yes, and since there is a crash
with that kernel anyway, its better
to just disable.

The very strange thing here is
that while esp is corrupted, the
code in dpmi.c doesn't crash
but the very same code in
msdoshlp.c - does.
Please do this:
```
diff --git a/src/dosext/dpmi/dpmi.c b/src/dosext/dpmi/dpmi.c
index 2bbbc0442..4f1cb95c1 100644
--- a/src/dosext/dpmi/dpmi.c
+++ b/src/dosext/dpmi/dpmi.c
@@ -2959,6 +2959,7 @@ static void make_iret_frame(sigcontext_t *scp, void *sp,
 {
   if (DPMI_CLIENT.is_32) {
     unsigned int *ssp = sp;
+D_printf("stack at %p\n", ssp);
     *--ssp = dpmi_flags_to_stack(_eflags);
     *--ssp = cs;
     *--ssp = eip;
diff --git a/src/dosext/dpmi/msdoshlp.c b/src/dosext/dpmi/msdoshlp.c
index 75d203208..0349923a0 100644
--- a/src/dosext/dpmi/msdoshlp.c
+++ b/src/dosext/dpmi/msdoshlp.c
@@ -112,6 +112,7 @@ static void do_dpmi_iret(sigcontext_t *scp)
     void *sp = SEL_ADR_CLNT(_ss, _esp, is_32);
     if (is_32) {
        unsigned int *ssp = sp;
+D_printf("stack2 at %p\n", ssp);
        _eip = *ssp++;
        _cs = *ssp++;
        _eflags = dpmi_flags_from_stack_iret(scp, *ssp++);
```

**Comment by ecm-pushbx** [comment]
> > What about WANT_SIGRETURN_WA 1 by the way? Do I need it because of my weird kernel version?
> 
> Yes, and since there is a crash with that kernel anyway, its better to just disable.

What do you mean by this? Disable what?
 
> The very strange thing here is that while esp is corrupted, the code in dpmi.c doesn't crash but the very same code in msdoshlp.c - does. Please do this:
> 
> ```
> diff --git a/src/dosext/dpmi/dpmi.c b/src/dosext/dpmi/dpmi.c
> index 2bbbc0442..4f1cb95c1 100644
> --- a/src/dosext/dpmi/dpmi.c
> +++ b/src/dosext/dpmi/dpmi.c
> @@ -2959,6 +2959,7 @@ static void make_iret_frame(sigcontext_t *scp, void *sp,
>  {
>    if (DPMI_CLIENT.is_32) {
>      unsigned int *ssp = sp;
> +D_printf("stack at %p\n", ssp);
>      *--ssp = dpmi_flags_to_stack(_eflags);
>      *--ssp = cs;
>      *--ssp = eip;
> diff --git a/src/dosext/dpmi/msdoshlp.c b/src/dosext/dpmi/msdoshlp.c
> index 75d203208..0349923a0 100644
> --- a/src/dosext/dpmi/msdoshlp.c
> +++ b/src/dosext/dpmi/msdoshlp.c
> @@ -112,6 +112,7 @@ static void do_dpmi_iret(sigcontext_t *scp)
>      void *sp = SEL_ADR_CLNT(_ss, _esp, is_32);
>      if (is_32) {
>         unsigned int *ssp = sp;
> +D_printf("stack2 at %p\n", ssp);
>         _eip = *ssp++;
>         _cs = *ssp++;
>         _eflags = dpmi_flags_from_stack_iret(scp, *ssp++);
> ```

Crashes after a while as usual. Output at https://pushbx.org/ecm/test/20220422/1110.txt and log with `-D9+M` at https://pushbx.org/ecm/test/20220422/1110.log (This directory also contains the current debugger executable and listing and map file.)

**Comment by stsp** [comment]
Please add also this:
```
@@ -186,7 +187,7 @@ static void iret2far(int tid, void *arg, void *arg2)
 {
     sigcontext_t *scp = arg2;
     struct pmaddr_s pma;
-
+D_printf("iret2far %s", DPMI_show_state(scp));
     pma.selector = _cs;
     pma.offset = _eip;
     coopth_push_user_data(tid, (void *)(uintptr_t)_eflags);
```

**Comment by ecm-pushbx** [comment]
Output: https://pushbx.org/ecm/test/20220422/1920.txt Log: https://pushbx.org/ecm/test/20220422/1920.log

**Comment by stsp** [comment]
OK, that's crazy, I still don't
see a problem...
Please add this:
```
diff --git a/src/dosext/dpmi/dpmi.c b/src/dosext/dpmi/dpmi.c
index 1bfa68d4f..5fea2fc8b 100644
--- a/src/dosext/dpmi/dpmi.c
+++ b/src/dosext/dpmi/dpmi.c
@@ -245,6 +245,7 @@ static void *SEL_ADR_LDT(unsigned short sel, unsigned int reg, int is_32)
     p = GetSegmentBase(sel) + reg;
   else
     p = GetSegmentBase(sel) + LO_WORD(reg);
+D_printf("sel_adr: %x %x %x\n", sel, reg, p);
   /* The address needs to wrap, also in 64-bit! */
   return LINEAR2UNIX(p);
 }
```

**Comment by ecm-pushbx** [comment]
I changed the last `%x` to `%lx` to avoid a compiler warning about a mismatching format.

Output: https://pushbx.org/ecm/test/20220422/2420.txt Log: https://pushbx.org/ecm/test/20220422/2420.log

**Comment by stsp** [comment]
Thanks, I think this is now
"fixed". At least to the best
we can have when espfix64
doesn't work properly.

**Comment by ecm-pushbx** [comment]
Has been running at the PM lDebugX prompt for over 40 minutes now without a crash. So it seems to be fixed. Thanks a lot!

Is it possible to add a configure step to automatically set `WANT_SIGRETURN_WA 1`?

**Comment by stsp** [comment]
I re-enabled it now and
also fixed the state printing.

**Comment by ecm-pushbx** [comment]
Thanks!
